File: generate_hash.js
// Генератор валидной подписи Telegram для тестирования
// Запустите в Node.js: node generate_hash.js

const crypto = require('crypto');

// Ваш BOT_TOKEN из .env
const BOT_TOKEN = "8265053392:AAE_D8DD1N9nR-KJ4Nq1mDEV3z5WN-qp6gk";

// Данные для теста
const testData = {
    "auth_date": "1234567890",
    "user": JSON.stringify({
        "id": 123456789,
        "username": "player",
        "first_name": "John"
    })
};

// Функция генерации подписи (как в Rust коде)
function generateTelegramHash(data, botToken) {
    // Сортируем ключи по алфавиту
    const sortedKeys = Object.keys(data).sort();
    
    // Формируем check_string (ключ=значение\n...)
    const checkString = sortedKeys
        .filter(key => key !== "hash")
        .map(key => `${key}=${data[key]}`)
        .join('\n');
    
    // Вычисляем секретный ключ (SHA256 от bot_token)
    const secretKey = crypto.createHash('sha256')
        .update(botToken)
        .digest();
    
    // Создаём HMAC-SHA256
    const hash = crypto.createHmac('sha256', secretKey)
        .update(checkString)
        .digest('hex');
    
    return hash;
}

const hash = generateTelegramHash(testData, BOT_TOKEN);

console.log('\n✅ Готовый JSON для Thunder Client:\n');
console.log(JSON.stringify({
    "auth_date": testData.auth_date,
    "user": JSON.parse(testData.user),
    "hash": hash
}, null, 2));

console.log('\n📝 Или в одну строку:\n');
console.log(JSON.stringify({
    "auth_date": testData.auth_date,
    "user": JSON.parse(testData.user),
    "hash": hash
}));

File: src/bin/generate_hash.rs
// Утилита для генерации валидной подписи Telegram для тестирования
// Запуск: cargo run --bin generate_hash

use hmac::{Hmac, Mac};
use sha2::{Sha256, Digest};
use hex;
use std::collections::BTreeMap;
use serde_json::json;

fn generate_telegram_hash(data: &BTreeMap<String, String>, bot_token: &str) -> String {
    let mut check_string = String::new();
    
    // Сортируем ключи в алфавитном порядке
    for (key, value) in data {
        if key != "hash" {
            check_string.push_str(&format!("{}={}\n", key, value));
        }
    }
    
    // Удаляем последний символ новой строки
    if check_string.ends_with('\n') {
        check_string.pop();
    }
    
    // Вычисляем секретный ключ
    let secret_key = Sha256::digest(bot_token.as_bytes());
    
    // Создаем HMAC
    let mut mac = Hmac::<Sha256>::new_from_slice(&secret_key)
        .expect("HMAC can take key of any size");
    mac.update(check_string.as_bytes());
    
    let result = mac.finalize().into_bytes();
    hex::encode(result)
}

fn main() {
    // Получаем BOT_TOKEN из env или используем из .env
    dotenvy::dotenv().ok();
    let bot_token = std::env::var("TELEGRAM_BOT_TOKEN")
        .expect("TELEGRAM_BOT_TOKEN должен быть в .env");
    
    // Тестовые данные
    // ВАЖНО: user должен быть сериализован в том же формате, как его обрабатывает сервер
    // Сервер использует value.to_string() для объектов, что даёт JSON без пробелов
    let user_json = serde_json::to_string(&json!({
        "id": 123456789,
        "username": "player",
        "first_name": "John"
    })).unwrap();
    
    let mut data = BTreeMap::new();
    data.insert("auth_date".to_string(), "1234567890".to_string());
    data.insert("user".to_string(), user_json);
    
    // Генерируем подпись
    let hash = generate_telegram_hash(&data, &bot_token);
    
    // Выводим готовый JSON для тестирования
    println!("\n✅ Готовый JSON для Thunder Client:\n");
    let json = json!({
        "auth_date": "1234567890",
        "user": {
            "id": 123456789,
            "username": "player",
            "first_name": "John"
        },
        "hash": hash
    });
    
    println!("{}", serde_json::to_string_pretty(&json).unwrap());
    
    println!("\n📝 Или в одну строку (для копирования):\n");
    println!("{}", serde_json::to_string(&json).unwrap());
}

